### PR TITLE
mesa: suport "all" in galliumDrivers and vulkanDrivers; purge enableOpenCL

### PR DESCRIPTION
**Pull request by PedroHLC**
## Description of changes

~This one didn't actually need to go to staging because it's not supposed to change any hash. But since 23.3 is not on master yet, I'll send it here.~ UPDATE: After applying K900 suggestions, this now changes the 32-bits Mesa's hash and can only target staging.

Instead of adding a conditional for NVK, we just normalized the dependencies with or without OpenCL.

PS: This file could use some formatting.

## Off-topics

24.0 will be more interesting with `video-codecs=all` and Meson-managed cargo-deps for NVK, but that's something to worry about when the time comes.

## Things done

Checked if it eval. (But building it, I hope it finishes before 2024)

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/mesa/default.nix [hunk 1]
```diff
@@ -233,20 +238,20 @@ self = stdenv.mkDerivation {
     python3Packages.python # for shebang
   ] ++ lib.optionals haveWayland [ wayland wayland-protocols ]
     ++ lib.optionals stdenv.isLinux [ libomxil-bellagio libva-minimal udev lm_sensors ]
-    ++ lib.optionals enableOpenCL [ llvmPackages.libclc llvmPackages.clang llvmPackages.clang-unwrapped spirv-llvm-translator ]
+    ++ lib.optionals needsRust [ llvmPackages.libclc llvmPackages.clang llvmPackages.clang-unwrapped spirv-llvm-translator ]
```
Comment (K900) [inline pkgs/development/libraries/mesa/default.nix:241]: Wrong conditional?

File: pkgs/development/libraries/mesa/default.nix [hunk 2]
```diff
@@ -214,9 +218,10 @@ self = stdenv.mkDerivation {
     # Clover, old OpenCL frontend
     "-Dgallium-opencl=icd"
     "-Dopencl-spirv=true"
-
     # Rusticl, new OpenCL frontend
-    "-Dgallium-rusticl=true" "-Drust_std=2021"
```
Comment (K900) [inline pkgs/development/libraries/mesa/default.nix:219]: Rusticl always needs-rust

File: pkgs/development/libraries/mesa/default.nix [hunk 3]
```diff
@@ -233,20 +238,20 @@ self = stdenv.mkDerivation {
...
     ++ lib.optionals stdenv.isLinux [ libomxil-bellagio libva-minimal udev lm_sensors ]
-    ++ lib.optionals enableOpenCL [ llvmPackages.libclc llvmPackages.clang llvmPackages.clang-unwrapped spirv-llvm-translator ]
+    ++ lib.optionals needsRust [ llvmPackages.libclc llvmPackages.clang llvmPackages.clang-unwrapped spirv-llvm-translator ]
     ++ lib.optional withValgrind valgrind-light
     ++ lib.optional haveZink vulkan-loader
     ++ lib.optional haveDozen directx-headers;
 
   depsBuildBuild = [ pkg-config ]
-    ++ lib.optional enableOpenCL buildPackages.stdenv.cc;
+    ++ lib.optional needsRust buildPackages.stdenv.cc;
```
Comment (K900) [inline pkgs/development/libraries/mesa/default.nix:247]: Wrong conditional here too?

File: pkgs/development/libraries/mesa/default.nix (hunk 1 above)
Comment (PedroHLC) [reply pkgs/development/libraries/mesa/default.nix:241]: Isn't `rusticl` enabled with `enableOpenCL`?

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:241]: It's not the only thing enabled with enableOpenCL. There's also the older Clover frontend.

File: pkgs/development/libraries/mesa/default.nix (hunk 2 above)
Comment (PedroHLC) [reply pkgs/development/libraries/mesa/default.nix:219]: It's logically impossible to have `enableOpenCL` without `needsRust` here.

File: pkgs/development/libraries/mesa/default.nix (hunk 1 above)
Comment (PedroHLC) [reply pkgs/development/libraries/mesa/default.nix:241]: NVK and Rusticl needs Rustc and `-Drust_std=2021`
NVK and Clover need libclc
:eyes:

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:241]: Then maybe we should just give up on making this stuff conditional?

Comment (PedroHLC) [reply pkgs/development/libraries/mesa/default.nix:241]: We will want these deps for the 32-bits NVK, but (at least right now) we don't build Clover or Rusticl in our 32-bits mesa.

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:241]: Maybe we should do that too.

Comment (PedroHLC) [reply pkgs/development/libraries/mesa/default.nix:241]: So options are:
1. Weird conditional that mix Rust and LLVM;
2. Always pull all deps (now including Rust, Clang, LibCLC, and SPIR-V) but keep Clover and Rusticl meson-options switchable;
3. Pull all deps and build the same stuff in all architectures.

Comment (K900) [reply pkgs/development/libraries/mesa/default.nix:241]: I'm leaning towards 3, if people really want to turn it off, they can always override mesonFlags.

**Comment by jansol** [comment]
Regarding OpenCL, [Clover is slated for deletion](https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/19385) and Ubuntu for example no longer ships it. Should we follow suit and not build clover by default? I'm not sure if clover is actually usable in the first place as I never got it to work. Rusticl still requires `RUSTICL_ENABLE=iris,radeonsi,r600` depending on your GPU, though...

**Comment by K900** [comment]
It is kind of usable and it doesn't really cost us anything to ship it until it's replaced upstream. I don't see why we would choose not to, really.

**Comment by jansol** [comment]
Is it? In that case, sure let's keep it. For me it has just always been broken and cluttering up the ICD loader.

**Comment by jansol** [comment]
Is there anything holding this up? Also tangentially related, [23.3.2 is out now](https://lists.freedesktop.org/archives/mesa-dev/2023-December/226088.html).

**Comment by K900** [comment]
FWIW I'm still not convinced `all` is a good idea, because at least on older versions it didn't actually enable _all_ the drivers.

**Comment by PedroHLC** [comment]
> not convinced `all` is a good idea

This PR does not change our derivation to use `"all"`, it allows one to use `["all"]` when overriding, without needing to do something as weird as `["all" "zink" "d3d12"]`, the other option would be just to document this uncommon solution.

**Comment by K900** [comment]
Yeah, but do we want to actually allow that? Given it's not the thing you generally want OR the thing you'd expect given the name.

**Comment by jansol** [comment]
So it's basically just changing the semantics of `"all"` where `["all" "zink" "d3d12"]` would be the upstream semantics?

I'd generally lean towards sticking with upstream semantics since that is much easier to discover and follow when reading the derivation.

**Comment by PedroHLC** [comment]
> So it's basically just changing the semantics of `"all"` where `["all" "zink" "d3d12"]` would be the upstream semantics?

We have specific code-paths that only enables when we have "zink" and "d3d12" in the list, even though "all" also enables "zink" and "d3d12", so right now we're breaking upstream behaviour.

**Comment by PedroHLC** [comment]
We can revisit these two changes individually with Mesa 24.0

**Pull request closed without merging**